lec02-basic-r.Rmd

```diff
@@ -1,4 +1,757 @@
+could also assign to a variable in the function and return the function.
+
+```{r}
+add_two_numbers <- function(num1, num2) {
+    my_sum <- num1 + num2
+    return(my_sum)
+}
+add_two_numbers(4, 5)
+```
+
+> **Challenge**
+>
+> Can you write a function that calculates the mean of 3 numbers?
+
+> ```{r}
+> mean_of_three_numbers <- function(num1, num2, num3) {
+>    my_sum <- num1 + num2 + num3
+>    my_mean <- my_sum / 3
+>    return(my_mean)
+> }
+> mean_of_three_numbers(2, 4, 6)
+> ```
+
+
+## Loops and vectorization
+
+Loop, or for-loops, are essential to programming in general. However, in R, you
+should avoid them as often as possible because there are more efficient ways of
+doing things that you should use instead. It is still important that you
+understand the concept of loops and you might also use them in some of your own
+functions if there is no vectorized way of going about what you want to do.
+
+You can think of a for-loop as: "for each number contained in a list/vector,
+perform this operation" and the syntax basically says the same thing:
+
+```{r}
+v <- c(2, 4, 6)
+for (num in v) {
+    print(num)
+}
+```
+
+Instead of printing out every number to the console, we could also add numbers
+cumulatively, to calculate the sum of all the numbers in the vector:
+
+```{r}
+# To increment `w` each time, we must first create the variable,
+# which we do by setting `w <- 0`, referred to as initializing.
+# This also ensures that `w` is zero at the start of the loop and
+# doesn't retain the value from last time we ran this code.
+w <- 0
+for (num in v) {
+    w <- w + num
+}
+w
+```
+
+If we put what we just did inside a function, we have essentially recreated the
+`sum` function in R.
+
+```{r}
+my_sum <- function(input_vector) {
+    vector_sum <- 0
+    for (num in input_vector){
+        vector_sum <- vector_sum + num
+    }
+    return(vector_sum)
+}
+
+my_sum(v)
+```
+
+Although this gives us the same output as the built-in function `sum`, the
+built-in function has many more optimizations so it is much faster than our
+function. In R, it is always faster to try to find a way of doing things without
+writing a loop yourself. When you are reading about R, you might see suggestions
+that you should try to *vectorize* your code to make it faster. What people are
+referring to, is that you should not write for loops in R and instead use the
+ready-made functions that are much more efficient in working with vectors and
+essentially performs operations on entire vector at once instead of
+one number at a time. If anyone is interested in more details about how this
+works, please ask me after class, but conceptually loops operate on one element
+at a time while vectorized code operate on all elements of a vector at once.
+
+*Go to the output options and set it to `html_document`, and then knit*
+
+**Remember to save your notebook file (.Rmd) on a USB stick or email it to
+yourself**
```

Comment:
These two lines can be deleted. They aren't necessary for the online lesson material, since this is a class/situation dependent issue.